### ISSUE TITLE
LSTM operator not present in the onnx model after conversion when using MultiRNNCell

### ISSUE THREAD

**Issue by sonu1-p** [onnx/tensorflow-onnx#311]
Hi. I am trying to restore a LSTM model from a tensorflow model checkpoint and convert it to onnx. The conversion is succeeding but the generated model does not have lstm operator.

Here are the details about the model -

input = tf.placeholder(tf.float32, [None, None, 192], name="input")
with tf.variable_scope("model", reuse=tf.AUTO_REUSE) as vs:
    lstm_cells = tf.contrib.rnn.MultiRNNCell([tf.contrib.rnn.LSTMCell(num_units) for _ in range(num_layers)])
    (output, final_state) = tf.nn.dynamic_rnn(cell=_lstm_cells, inputs=input, dtype=tf.float32, time_major=True)
    output = tf.identity(output, name="my_output")
    saver = tf.train.Saver()
    sess = tf.Session()
    saver.restore(sess,'/Users/aluma/tensorflow-models/model.ckpt')

    # Freeze the graph
    min_graph = tf.compat.v1.graph_util.convert_variables_to_constants(sess, sess.graph_def, ["model/my_output"])
    tf.compat.v1.graph_util.remove_training_nodes(min_graph)
    tf.train.write_graph(min_graph, ".", "frozen_final.pb", as_text=False)

Then converting it to onnx as:
python -m tf2onnx.convert --input frozen_final.pb --inputs input:0 --outputs model/my_output:0 --output model.onnx --verbose

tensorflow ops: Counter({'Const': 109, 'Identity': 28, 'Enter': 28, 'ConcatV2': 19, 'Sigmoid': 15, 'Mul': 15, 'Merge': 13, 'Switch': 13, 'NextIteration': 13, 'Add': 12, 'ExpandDims': 10, 'Fill': 10, 'Tanh': 10, 'MatMul': 7, 'BiasAdd': 7, 'Reshape': 6, 'Shape': 5, 'Split': 5, 'GatherV2': 4, 'Prod': 4, 'Transpose': 4, 'StridedSlice': 3, 'Pack': 2, 'TensorArrayV3': 2, 'Range': 2, 'Less': 2, 'Placeholder': 1, 'TensorArrayScatterV3': 1, 'Maximum': 1, 'Minimum': 1, 'LogicalAnd': 1, 'LoopCond': 1, 'TensorArrayReadV3': 1, 'TensorArrayWriteV3': 1, 'Exit': 1, 'TensorArraySizeV3': 1, 'TensorArrayGatherV3': 1})
tensorflow attr: Counter({'T': 236, 'dtype': 114, 'value': 109, 'N': 34, 'is_constant': 28, 'parallel_iterations': 28, 'frame_name': 28, '_class': 25, 'Tidx': 25, 'Tdim': 10, 'index_type': 10, 'transpose_a': 7, 'transpose_b': 7, 'data_format': 7, 'Tshape': 6, 'out_type': 5, 'num_split': 5, 'Tindices': 4, 'Tparams': 4, 'Taxis': 4, 'keep_dims': 4, 'Tperm': 4, 'Index': 3, 'shrink_axis_mask': 3, 'ellipsis_mask': 3, 'begin_mask': 3, 'new_axis_mask': 3, 'end_mask': 3, 'element_shape': 3, 'axis': 2, 'clear_after_read': 2, 'dynamic_size': 2, 'identical_element_shapes': 2, 'tensor_array_name': 2, 'shape': 1})
onnx mapped: Counter({'Const': 96, 'ConcatV2': 19, 'Identity': 15, 'Sigmoid': 15, 'Mul': 15, 'Add': 11, 'ExpandDims': 10, 'Tanh': 10, 'Fill': 10, 'MatMul': 7, 'BiasAdd': 7, 'Reshape': 6, 'Split': 5, 'Transpose': 4, 'GatherV2': 4, 'Prod': 4, 'Shape': 3, 'Pack': 2, 'Placeholder': 1, 'StridedSlice': 1})
onnx unmapped: Counter()
DEBUG:tf2onnx.optimizer.transpose_optimizer:finish after 4 iteration(s)
INFO:tf2onnx.optimizer.transpose_optimizer: 0 transpose op(s) left, ops diff after transpose optimization: Counter({'Cast': 0, 'Unsqueeze': 0, 'Reshape': 0, 'Concat': 0, 'Shape': 0, 'Gather': 0, 'ReduceProd': 0, 'MatMul': 0, 'Add': 0, 'Slice': 0, 'Squeeze': 0, 'Tile': 0, 'Scan': 0, 'Identity': 0, 'Transpose': -4})

I enabled debug logging in tf2onnx to see why the LSTM node is not being produced. I see this in the logs:

DEBUG:tf2onnx.rewriter.rnn_unit_writer_base:=========================
DEBUG:tf2onnx.rewriter.rnn_unit_writer_base:find_dynamic_run_unit >> start handling a new potential rnn cell
DEBUG:tf2onnx.rewriter.rnn_unit_writer_base:rnn scope name is model/rnn/while/rnn/multi_rnn_cell
DEBUG:tf2onnx.rewriter.rnn_unit_writer_base:find_dynamic_run_unit >> get_weight_and_bias starts
DEBUG:tf2onnx.rewriter.rnn_utils:found weights model/rnn/multi_rnn_cell/cell_3/lstm_cell/kernel
DEBUG:tf2onnx.rewriter.rnn_utils:found weights model/rnn/multi_rnn_cell/cell_3/lstm_cell/bias
DEBUG:tf2onnx.rewriter.rnn_utils:found weights model/rnn/while/rnn/multi_rnn_cell/cell_3/lstm_cell/add/y
DEBUG:tf2onnx.rewriter.rnn_unit_writer_base:No LoopCond op is found, skip
DEBUG:tf2onnx.rewriter.rnn_unit_writer_base:no cell variable initializers found, skip

Because of no LoopCond in model/rnn/while/rnn/multi_rnn_cell scope t seems it not able to deduce that it is a lstm cell, is my understanding correct?

I see LoopCond but in this scope:
model/rnn/while/LogicalAnd
model/rnn/while/LoopCond
model/rnn/while/Switch

Is this something wrong with my model or the tf2onnx tool is not handling this conversion correctly? Please let me know if you need any more information from me.

**Comment by pengwa** [comment]
LSTM + MultiRNNCell will be converted to Scan op. Because ONNX LSTM capabilities won't cover this case. Please check the generated file for the Scan op.

**Comment by sonu1-p** [comment]
Thank you for the quick response pengwa! Yes, the generated model does have scan operator.

**Comment by sonu1-p** [comment]
Sorry, I am new to onnx and tensorflow. Can you please help me understand this onnx limitation which does not allow multiRnnCell's LSTM operators to be used as LSTMs?

**Comment by pengwa** [comment]
https://github.com/onnx/onnx/blob/master/docs/Operators.md#LSTM defined the LSTM operator, it assumed the calculation cell graph to be a fixed one. If you use Multiple layer of lstm in TF, the lstm cells are stacked together as a single and bigger cell. So ONNX LSTM won't work here.

**Comment by sonu1-p** [comment]
I see. Thank you so much for the explanation! Also, not related this, are there any plans to add cuDnnCompatibleLSTM support to this library?

**Comment by pengwa** [comment]
it's based on https://github.com/tensorflow/tensorflow/blob/3ffbbfb5a9e4cf0ae777453f0c70bc95c1e17d4e/tensorflow/contrib/rnn/python/ops/lstm_ops.py#L335. Essentially it is using a fused op within TF graph, so it would need additional logic doing this. We usually prioritize this with real models. If you have models in hande and want it converted, you can consider to add the support.

**Comment by sonu1-p** [comment]
Got it. Thank you!
